consumer.go

```diff
@@ -156,31 +164,34 @@ func (c *Consumer) Scan(ctx context.Context, fn func(*Record) bool) error {
 // ScanShard loops over records on a specific shard, calls the callback func
 // for each record and checkpoints the progress of scan.
 // Note: Returning `false` from the callback func will end the scan.
-func (c *Consumer) ScanShard(ctx context.Context, shardID string, fn func(*Record) bool) (err error) {
+func (c *Consumer) ScanShard(ctx context.Context, shardID string, fn func(*Record) ScanError) (err error) {
 	lastSeqNum, err := c.checkpoint.Get(c.streamName, shardID)
 	if err != nil {
 		return fmt.Errorf("get checkpoint error: %v", err)
 	}
 
 	c.logger.Println("scanning", shardID, lastSeqNum)
-
 	// get records
 	recc, errc, err := c.client.GetRecords(ctx, c.streamName, shardID, lastSeqNum)
 	if err != nil {
 		return fmt.Errorf("get records error: %v", err)
 	}
-
 	// loop records
 	for r := range recc {
-		if ok := fn(r); !ok {
+		scanError := fn(r)
+		if scanError.StopScan {
 			break
 		}
+		if scanError.Error != nil {
+			c.logger.Println(fmt.Errorf("record processing error: %v", scanError.Error))
+		}
```

Comment:
can remove the indentation of skip checkpoint

```go
for r := range recc {
    scanError := fn(r)
    
    if scanError.StopScan {
      break
    }

    if scanError.SkipCheckpoint {
      continue
    }

    c.counter.Add("records", 1)

    err := c.checkpoint.Set(c.streamName, shardID, *r.SequenceNumber)
    if err != nil {
      return fmt.Errorf("set checkpoint error: %v", err)
    }
}
```

also feels like there might be a logic issue between "stop" and "skip" i.e. maybe someone wants to stop, but also checkpoint?